crates/ironrdp-svc/src/lib.rs

```diff
@@ -11,17 +11,24 @@ use alloc::boxed::Box;
 use alloc::collections::BTreeMap;
 use core::any::{Any, TypeId};
 use core::fmt;
+use std::borrow::Cow;
 use std::marker::PhantomData;
 
 use bitflags::bitflags;
 use ironrdp_pdu::gcc::{ChannelName, ChannelOptions};
 use ironrdp_pdu::write_buf::WriteBuf;
-use ironrdp_pdu::{assert_obj_safe, PduResult};
+use ironrdp_pdu::{assert_obj_safe, mcs, PduResult};
 use pdu::cursor::WriteCursor;
 use pdu::gcc::ChannelDef;
 use pdu::rdp::vc::ChannelControlFlags;
 use pdu::{custom_err, encode_buf, PduEncode, PduParsing};
 
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum ChannelSide {
+    Server,
+    Client,
+}
```

Comment:
note: I’m okay with keeping this as a small utility, but it seems to me that this enum does not really need to be part of public API of `ironrdp-svc` and does not add much value.

rationale: `ironrdp-svc`  does not provide any interesting logic relying on this type. Consumers of `ironrdp-svc` don’t need to use it for anything. I could imagine many channels using two different types, one per side, instead of unifying both sides on a single type. I could also imagine a boolean being used. (The built-in type actually provides many interesting utility methods which could make it a superior alternative in some cases.)

suggestion: At least in this case, my preference goes for this type of API which does not require the library user to know about the type `ChannelSide` at all:
```rust
/// CLIPRDR static virtual channel client endpoint implementation
#[derive(Debug)]
pub struct Cliprdr {
    /* … */
    is_server: bool, // or: side: ChannelSide,
}

impl Cliprdr {
  pub fn client_new(backend: Box<dyn CliprdrBackend>) -> Self {
    Self {
      /* … */,
      is_server: false, // or: side: ChannelSide::Client,
    }
  }

  pub fn server_new(backend: Box<dyn CliprdrBackend>) -> Self {
    Self {
      /* … */,
      is_server: true, // or: side: ChannelSide::Server,
    }
  }
}
```

rationale: It’s typically not useful to flip the "server" flag in an isolated manner. Acting as a server (or as a client) requires many differences in the code itself, and I can’t imagine a good case for something like this:
```rust
let side = if act_as_a_server {
    ChannelSide::Server
} else {
    ChannelSide::Client
};

Cliprdr::new(Box::new(clipboard_backend), side);
```

Reply:
I agree with your feeling, but ironrdp-svc is also used to implement channels themself, and channel implementation usually need to know which side they act as. I am fine with using different constructuors if you prefer, but new(Client) or new_client() is kind of the same to me at this point.

Reply:
And channel that are asymmetricals don't have to use the ChannelSide in their public API. For ex, a GFX channel implementation could expose GFXClient & GFXServer types.

Reply:
Also, I agree with @CBenoit that adding `enum` to the constructor is not optimal. Also, this could result in a situation, where API allows to create, for example, server-side svc, but accidentally use client-side `ChannelProcessor`. Therefore I propose the following solution:

Move client/server decision to `ChannelProcessor` by introducing `is_server()` or `peer_kind()` method to this trait. That way, the svc channel could add correct headers based on the value returned by `ChannelProcessor`, while the `ChannelProcessor` implementors themselves are always will be aware of its peer kind (client/server).

I also suggest splitting `Cliprdr` to `CliprdrClient` and `CliprdrServer`, without introduction of PeerKind constructor argument - we could always re-use the same communication sequences by moving it to a shared module, and svc logic will be aware of the peer kind by returning `is_server() -> true` in `CliprdrServer` and `is_server() -> false`

Reply:
> I also suggest splitting Cliprdr to CliprdrClient and CliprdrServer, without introduction of PeerKind constructor argument - we could always re-use the same communication sequences by moving it to a shared module

I also prefer this approach. Less room for misuses.

> Move client/server decision to `ChannelProcessor` by introducing `is_server()` or `peer_kind()` method to this trait. That way, the svc channel could add correct headers based on the value returned by `ChannelProcessor`, while the `ChannelProcessor` implementors themselves are always will be aware of its peer kind (client/server).

I’m not sure to follow this part, I may be missing something.
To warrant a new method in a trait, the new method must serve a purpose in generic code.
Is the "peer kind" information _generically_ useful somewhere?
Do you think of a specific place specifically?
My understanding is that the only component which needs to know, is the static channel itself, and it already knows whether it is "client" or "server" depending on whether it is a `CliprdrClient` or `CliprdrServer` (or, depending on the approach taken, whether `client_new` was called instead of `server_new`, but I still prefer different types).

Reply:
@CBenoit  the rationale about adding new trait method is that `StaticVirtualChannel` is also need to know how to wrap svc messages inside MDUs PDUs (see logic [here](https://github.com/Devolutions/IronRDP/pull/333/files/c7ec00480270b70ab7b7c86e33bdbb13e1e543c4#diff-419140c5aaf75e62c1ccce87c9c3f5848a8180914fa4b9608e8a4eaba502a5acR180-R196)). So we can't leave `StaticVirtualChannel` without knowledge about underlying `ChannelProcessor` kind 😢

Reply:
> @CBenoit the rationale about adding new trait method is that `StaticVirtualChannel` is also need to know how to wrap svc messages inside MDUs PDUs (see logic [here](https://github.com/Devolutions/IronRDP/pull/333/files/c7ec00480270b70ab7b7c86e33bdbb13e1e543c4#diff-419140c5aaf75e62c1ccce87c9c3f5848a8180914fa4b9608e8a4eaba502a5acR180-R196)). So we can't leave `StaticVirtualChannel` without knowledge about underlying `ChannelProcessor` kind 😢

Oh, I see. Then, actually we don’t need to ask the processor for this information. We inherently know the side, because client code will call [client_encode_svc_messages](https://github.com/Devolutions/IronRDP/pull/333/files/c7ec00480270b70ab7b7c86e33bdbb13e1e543c4#diff-419140c5aaf75e62c1ccce87c9c3f5848a8180914fa4b9608e8a4eaba502a5acR207), and server code will call [server_encode_svc_messages](https://github.com/Devolutions/IronRDP/pull/333/files/c7ec00480270b70ab7b7c86e33bdbb13e1e543c4#diff-419140c5aaf75e62c1ccce87c9c3f5848a8180914fa4b9608e8a4eaba502a5acR217). This function is simply refactoring the logic, but could be two separate functions, without ifs. (Actually, that would be my preference, but it’s more of a nitpick than anything else.)

Reply:
I have limited the role with a trait specific to cliprdr, assuming the nearly symmetrical implementation of the channel is going to be an exception:
```
pub type CliprdrClient = Cliprdr<Client>;
pub type CliprdrServer = Cliprdr<Server>;

impl SvcClientProcessor for CliprdrClient {}
impl SvcServerProcessor for CliprdrServer {}
```

If you have better ideas, this could later be changed too.